### PR TITLE
snap-repair: implement `snap-repair {list,show}`

### PR DESCRIPTION
**Pull request by mvo5**
The `snap-repair list` output:
```
Issuer     Seq  Rev  Status  Summary
canonical  1    3    retry   repair one
my-brand   1    1    done    my-brand repair one
my-brand   2    2    skip    my-brand repair two
```

And `snap-repair show canonical-1`:
```
repair: canonical-1
status: retry
summary: repair one
rev: 3
script:
  #!/bin/sh
  echo retry output
output:
  retry output
```

Based on https://github.com/snapcore/snapd/pull/3777

### CODE REVIEW

File: cmd/snap-repair/cmd_show.go [hunk 1]
```diff
@@ -0,0 +1,87 @@
...
+	}
+	if len(repairTraces) == 0 {
+		return fmt.Errorf("cannot find repair \"%s-%s\"", brand, seq)
+	}
+
+	for _, trace := range repairTraces {
+		fmt.Fprintf(w, "repair: %s\n", trace.Repair())
+		fmt.Fprintf(w, "status: %s\n", trace.Status())
+		fmt.Fprintf(w, "summary: %s\n", trace.Summary())
+		fmt.Fprintf(w, "rev: %s\n", trace.Rev())
```
Comment (pedronis) [inline cmd/snap-repair/cmd_show.go]: I think we want "revision:"  here   (we never use the rev abbreviation for assertion revisions externally)

File: cmd/snap-repair/trace.go [hunk 2]
```diff
@@ -0,0 +1,179 @@
...
+// Repair returns the repair human readable string in the form $brand-$id
+func (rt *repairTrace) Repair() string {
+	seq := filepath.Base(filepath.Dir(rt.path))
+	brand := filepath.Base(filepath.Dir(filepath.Dir(rt.path)))
+
+	return fmt.Sprintf("%s-%s", brand, seq)
+}
+
+// Rev returns the revision of the repair
+func (rt *repairTrace) Rev() string {
```
Comment (pedronis) [inline cmd/snap-repair/trace.go]: let's call this Revision

File: cmd/snap-repair/trace.go [hunk 3]
```diff
@@ -0,0 +1,179 @@
...
+	f, err := os.Open(rt.path)
+	if err != nil {
+		return "cannot read summary"
+	}
+	defer f.Close()
+
+	needle := "summary: "
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		s := scanner.Text()
```
Comment (pedronis) [inline cmd/snap-repair/trace.go]: this assumes that summary has no newlines in it, that's reasonable but we need to enforce it then in asserts/repair.go

File: cmd/snap-repair/trace.go [hunk 4]
```diff
@@ -0,0 +1,179 @@
...
+
+	return "cannot find summary"
+}
+
+// Status returns the status of the given repair {done,skip,retry,running}
+func (rt *repairTrace) Status() string {
+	return filepath.Ext(rt.path)[1:]
+}
+
+func indentPrefix(level int) string {
```
Comment (pedronis) [inline cmd/snap-repair/trace.go]: isn't this   strings.Repeat(" ", level) ?

File: cmd/snap-repair/cmd_show.go (hunk 1 above)
Comment (pedronis) [reply cmd/snap-repair/cmd_show.go]: also revision should come after "repair" and before "status" I think

File: cmd/snap-repair/trace.go (hunk 4 above)
Comment (mvo5) [reply cmd/snap-repair/trace.go]: \o/

File: cmd/snap-repair/cmd_list.go [hunk 5]
```diff
@@ -0,0 +1,74 @@
...
+	repairTraces, err := newRepairTraces("*", "*")
+	if err != nil {
+		return err
+	}
+	if len(repairTraces) == 0 {
+		fmt.Fprintf(w, "no repairs yet\n")
+		return nil
+	}
+
+	fmt.Fprintf(w, "Repair\tRevision\tStatus\tSummary\n")
```
Comment (pedronis) [inline cmd/snap-repair/cmd_list.go]: I think Rev here (to save space for summary) was fine, we do the same in 'snap list' for example

File: cmd/snap-repair/trace.go [hunk 6]
```diff
@@ -0,0 +1,175 @@
...
+		return
+	}
+	defer f.Close()
+
+	scanner := bufio.NewScanner(f)
+	for scanner.Scan() {
+		fmt.Fprintf(w, "%s%s\n", indentPrefix(indent), scanner.Text())
+	}
+	if scanner.Err() != nil {
+		fmt.Fprintf(w, "%serror: %s\n", indentPrefix(indent), scanner.Err())
```
Comment (niemeyer) [inline cmd/snap-repair/trace.go]: Same.

File: cmd/snap-repair/trace.go [hunk 7]
```diff
@@ -0,0 +1,175 @@
...
+	}
+}
+
+// WriteOutputIndented outputs the repair output to the given writer w
+// with the indent level given by indent.
+func (rt *repairTrace) WriteOutputIndented(w io.Writer, indent int) {
+	f, err := os.Open(rt.path)
+	if err != nil {
+		fmt.Fprintf(w, "  error: %s\n", err)
```
Comment (niemeyer) [inline cmd/snap-repair/trace.go]: Same here and below.

File: cmd/snap-repair/trace.go [hunk 8]
```diff
@@ -0,0 +1,175 @@
...
+func (rt *repairTrace) validSuffix(traceName string) bool {
+	for _, valid := range []string{".retry", ".skip", ".done", ".running"} {
+		if strings.HasSuffix(traceName, valid) {
+			return true
+		}
+	}
+	return false
+}
+
+// revFromFilepath is a helper that extrace the revision number from the
```
Comment (niemeyer) [inline cmd/snap-repair/trace.go]: extrace? I like the word.. :)

File: cmd/snap-repair/trace.go [hunk 9]
```diff
@@ -0,0 +1,175 @@
...
+// filename of the repairTrace
+func revFromFilepath(name string) string {
+	var rev int
+	if _, err := fmt.Sscanf(filepath.Base(name), "r%d.", &rev); err == nil {
+		return strconv.Itoa(rev)
+	}
+	return "?"
```
Comment (niemeyer) [inline cmd/snap-repair/trace.go]: This sounds surprising and I'm afraid of ending up with "?" in curious and potentially dangerous places. Similar to above, I think it should be responsibility of the call site to decide what to do when it goes wrong.

File: cmd/snap-repair/cmd_show.go [hunk 10]
```diff
@@ -0,0 +1,87 @@
...
+
+	repairTraces, err := newRepairTraces(brand, seq)
+	if err != nil {
+		return err
+	}
+	if len(repairTraces) == 0 {
+		return fmt.Errorf("cannot find repair \"%s-%s\"", brand, seq)
+	}
+
+	for _, trace := range repairTraces {
```
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/snap-repair/cmd_show.go]: Yeah, I think the `---` is nice here. Right now we show multiple revisions but we make it not very clear that they belong together.

File: cmd/snap-repair/trace.go [hunk 11]
```diff
@@ -0,0 +1,175 @@
...
+// Revision returns the revision of the repair
+func (rt *repairTrace) Revision() string {
+	return revFromFilepath(rt.path)
+}
+
+// Summary returns the summary of the repair that was run
+func (rt *repairTrace) Summary() string {
+	f, err := os.Open(rt.path)
+	if err != nil {
+		return "cannot read summary"
```
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/snap-repair/trace.go]: Sure, fixed.

File: cmd/snap-repair/trace.go [hunk 12]
```diff
@@ -0,0 +1,175 @@
...
+	return strings.Repeat(" ", level)
+}
+
+// WriteScriptIndented outputs the script that produced this repair output
+// to the given writer w with the indent level given by indent.
+func (rt *repairTrace) WriteScriptIndented(w io.Writer, indent int) {
+	scriptPath := rt.path[:strings.LastIndex(rt.path, ".")] + ".script"
+	f, err := os.Open(scriptPath)
+	if err != nil {
+		fmt.Fprintf(w, "cannot read script: %v", err)
```
[the remark this answers is not in the document]
Comment (mvo5) [reply cmd/snap-repair/trace.go]: Indeed, I fixed this now.

File: cmd/snap-repair/trace.go (hunk 8 above)
Comment (mvo5) [reply cmd/snap-repair/trace.go]: This is why I'm member in https://launchpad.net/~blatant-and-awkward ;)

Comment (mvo5) [reply cmd/snap-repair/trace.go]: Fixed.

File: cmd/snap-repair/trace.go [hunk 13]
```diff
@@ -0,0 +1,176 @@
...
+		if scanner.Text() == "output:" {
+			break
+		}
+	}
+	// write the script output to w
+	for scanner.Scan() {
+		fmt.Fprintf(w, "%s%s\n", indentPrefix(indent), scanner.Text())
+	}
+	if scanner.Err() != nil {
+		return err
```
Comment (pedronis) [inline cmd/snap-repair/trace.go]: this should be return scanner.Err() ?

File: cmd/snap-repair/trace.go (hunk 6 above)
Comment (pedronis) [reply cmd/snap-repair/trace.go]: this should be  return scanner.Err()

File: cmd/snap-repair/cmd_show.go [hunk 14]
```diff
@@ -0,0 +1,89 @@
...
+		fmt.Fprintf(w, "status: %s\n", trace.Status())
+		fmt.Fprintf(w, "summary: %s\n", trace.Summary())
+
+		fmt.Fprintf(w, "script:\n")
+		if err := trace.WriteScriptIndented(w, 2); err != nil {
+			fmt.Fprintf(w, "%serror: %s", indentPrefix(2), err)
+		}
+
+		fmt.Fprintf(w, "output:\n")
+		trace.WriteOutputIndented(w, 2)
```
Comment (pedronis) [inline cmd/snap-repair/cmd_show.go]: this now need to handle error from WriteOutputIndented like for WriteScriptIndented

**Comment by pedronis** [comment]
last commit here needs to be backed out

**Comment by pedronis** [comment]
now that I merged the expose 'repair' PR this probably needs another master merge, and maybe a test fix? TestRepairHasCorrectPath is failing in some autopkgtests (not because of timing)

**Comment by pedronis** [comment]
@mvo5  as we chatted some kind of execution time recorded and shown by 'snap repair' would be nice, can be done here or in a follow up. ATM still is failing spread :/

**Comment by mvo5** [comment]
I will do the execution time in a followup.

**Pull request merged**